cpp/src/io/json/reader_impl.cu

```diff
@@ -603,52 +558,65 @@ reader::impl::impl(std::vector<std::unique_ptr<datasource>>&& sources,
  *
  * @return Table and its metadata
  */
-table_with_metadata reader::impl::read(json_reader_options const& options,
-                                       rmm::cuda_stream_view stream)
+table_with_metadata read_json(std::vector<std::unique_ptr<datasource>>& sources,
+                              json_reader_options const& reader_opts,
+                              rmm::cuda_stream_view stream,
+                              rmm::mr::device_memory_resource* mr)
 {
-  auto range_offset      = options.get_byte_range_offset();
-  auto range_size        = options.get_byte_range_size();
-  auto range_size_padded = options.get_byte_range_size_with_padding();
+  CUDF_EXPECTS(not sources.empty(), "No sources were defined");
 
-  ingest_raw_input(range_offset, range_size, range_size_padded);
-  CUDF_EXPECTS(buffer_.size() != 0, "Ingest failed: input data is null.\n");
+  CUDF_EXPECTS(reader_opts.is_enabled_lines(), "Only JSON Lines format is currently supported.\n");
 
-  decompress_input(stream);
-  CUDF_EXPECTS(uncomp_data_ != nullptr, "Ingest failed: uncompressed input data is null.\n");
-  CUDF_EXPECTS(uncomp_size_ != 0, "Ingest failed: uncompressed input data has zero size.\n");
+  auto parse_opts = parse_options{',', '\n', '\"', '.'};
 
-  auto rec_starts = find_record_starts(stream);
-  CUDF_EXPECTS(!rec_starts.is_empty(), "Error enumerating records.\n");
+  parse_opts.trie_true  = cudf::detail::create_serialized_trie({"true"}, stream);
+  parse_opts.trie_false = cudf::detail::create_serialized_trie({"false"}, stream);
+  parse_opts.trie_na    = cudf::detail::create_serialized_trie({"", "null"}, stream);
 
-  upload_data_to_device(rec_starts, stream);
-  CUDF_EXPECTS(data_.size() != 0, "Error uploading input data to the GPU.\n");
+  parse_opts.dayfirst = reader_opts.is_enabled_dayfirst();
 
-  set_column_names(rec_starts, stream);
-  CUDF_EXPECTS(!metadata_.column_names.empty(), "Error determining column names.\n");
+  auto range_offset      = reader_opts.get_byte_range_offset();
```

Comment:
At a future date, would there be value in a user-defined conversion (a `parse_options` constructor perhaps), to convert from `json_reader_options` to `parse_options`? There seem to be a fair few lines dedicated to setting up the `parse_options` here.